build/provenance.go

```diff
@@ -0,0 +1,198 @@
+		} else if err != nil {
+			return nil, err
+		}
+		if ev.Record == nil {
+			continue
+		}
+		if ev.Record.Result != nil {
+			provenanceDgst := provenanceDigest(ev.Record.Result)
+			if provenanceDgst == nil {
+				continue
+			}
+			eg.Go(func() error {
+				dt, err := getBlob(ctx, c, *provenanceDgst, "", l)
+				if err != nil {
+					return errors.Wrapf(err, "failed to load provenance from build record")
+				}
+				prv, err := encodeProvenance(dt, mode)
+				if err != nil {
+					return err
+				}
+				mu.Lock()
+				if out == nil {
+					out = make(map[string]string)
+				}
+				out["buildx.build.provenance"] = prv
+				mu.Unlock()
+				return nil
+			})
+		} else if ev.Record.Results != nil {
+			for platform, res := range ev.Record.Results {
+				platform := platform
+				provenanceDgst := provenanceDigest(res)
+				if provenanceDgst == nil {
+					continue
+				}
+				eg.Go(func() error {
+					dt, err := getBlob(ctx, c, *provenanceDgst, platform, l)
+					if err != nil {
+						return errors.Wrapf(err, "failed to load provenance from build record")
+					}
+					prv, err := encodeProvenance(dt, mode)
+					if err != nil {
+						return err
+					}
+					mu.Lock()
+					if out == nil {
+						out = make(map[string]string)
+					}
+					out["buildx.build.provenance/"+platform] = prv
+					mu.Unlock()
+					return nil
+				})
+			}
+		}
+	}
+	return out, eg.Wait()
+}
+
+func provenanceDigest(res *controlapi.BuildResultInfo) *digest.Digest {
+	for _, a := range res.Attestations {
+		if a.MediaType == "application/vnd.in-toto+json" && strings.HasPrefix(a.Annotations["in-toto.io/predicate-type"], "https://slsa.dev/provenance/") {
+			return &a.Digest
+		}
+	}
+	return nil
+}
+
+func encodeProvenance(dt []byte, mode confutil.MetadataProvenanceMode) (string, error) {
+	if mode == confutil.MetadataProvenanceModeMax {
+		return base64.StdEncoding.EncodeToString(dt), nil
+	}
+
+	var prv provenancetypes.ProvenancePredicate
+	if err := json.Unmarshal(dt, &prv); err != nil {
+		return "", errors.Wrapf(err, "failed to unmarshal provenance")
+	}
+
+	// remove buildConfig and metadata for minimal provenance
+	prv.BuildConfig = nil
+	prv.Metadata = nil
+
+	dtmin, err := json.Marshal(prv)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to marshal minimal provenance")
+	}
+
+	return base64.StdEncoding.EncodeToString(dtmin), nil
+}
+
+func getBlob(ctx context.Context, c *client.Client, dgst digest.Digest, platform string, l progress.SubLogger) (dt []byte, err error) {
+	id := "fetching " + dgst.String()
+	if platform != "" {
+		id = fmt.Sprintf("[%s] fetching %s", platform, dgst.String())
+	}
+	st := &client.VertexStatus{
+		ID: id,
+	}
+
+	defer func() {
+		now := time.Now()
+		st.Completed = &now
+		if err == nil {
+			st.Total = st.Current
+		}
+		l.SetStatus(st)
+	}()
+
+	now := time.Now()
+	st.Started = &now
+	l.SetStatus(st)
+
+	store := proxy.NewContentStore(c.ContentClient())
+	ra, err := store.ReaderAt(ctx, ocispecs.Descriptor{
```

Comment:
Assuming we can drop the logging, `content.ReadBlob` helper can be used for this.